### DISCUSSION TITLE
Output text to bottom screen in debug mode?

### DISCUSSION THREAD

**Discussion by Arkham-Slayer** [profi200/open_agb_firm#102] [Q&A]
I'm new to the 3ds homebrew development scene. I am trying to see if I can sort of have the X button act as B without un-mapping B in open_agb_firm like how gba vc inject has B to both B and X (I have already changed screen adjustment hotkey to Y as X would affect my new mapping). Specifically I am trying to get open_agb_firm to think B is being pushed when X is pressed.

I'm not sure if it's even reading my code since I have created a new function for this, so I am trying to have it output to the touch screen in the debug build since other outputs go there as well. From what I could  understand of the source code I thought ee_puts("B has been pressed") would be the print function I need but it's not? Is there any function to print to the touch screen like the way I am proposing? Is what I am trying to do overall already tried by someone else and just isn't feasible?

Thanks in advance!

**Comment by profi200** [comment]
ee_puts() prints a string to the console on the bottom screen, yes. And ee_printf() can also output numbers. It's not exactly clear to me if you just want to make the code do something when pressing X or you want to remap X so it acts as B in games.

**Reply by Arkham-Slayer** [reply]
I phrased it weirdly. I do want to "remap" X to B without changing B. As I stated GBA VC injects handle controls like this and I personally prefer it this way as I can keep default A and B for menu navigation and jrpgs, while also having attack (B) being top-left from jump (A) in action games almost snes-style.


The issue with the code is that my new function to handle this is not being called on. When I get back, I'm gonna try calling it from the main and see if that works. Thank you again.

**Reply by profi200** [reply]
That's unfortunately a bit more complicated. Take a look [here](https://github.com/profi200/libn3ds/blob/0701508ad76a80216244b8db2d73fb3685a5678e/source/arm11/drivers/lgy.c#L163) how input remapping is done. The pad_val and pad_sel registers are what you want. Note that remapping can introduce one extra frame of input delay depending on when the game reads the inputs. This is one reason why i didn't bother implementing button remapping yet.

Also see https://www.3dbrew.org/wiki/PDN_Registers#LGY_HIDEMU_MASK (and LGY_HIDEMU_PAD under it) for how they work. It's just different names for the same registers.
